controller/volume_controller.go

```diff
@@ -2688,11 +2688,59 @@ func (c *VolumeController) upgradeEngineForVolume(v *longhorn.Volume, es map[str
 		return nil
 	}
 
-	// only start live upgrade if volume is healthy
 	if v.Status.State != longhorn.VolumeStateAttached || v.Status.Robustness != longhorn.VolumeRobustnessHealthy {
+		if v.Status.State != longhorn.VolumeStateAttached || v.Status.Robustness != longhorn.VolumeRobustnessDegraded {
+			return nil
+		}
+		// Clean up inactive replica corresponding to non-existing active replica.
+		// This will allow volume controller to schedule and rebuild a new active replica
+		// and make the volume healthy again for the live upgrade to be able to continue.
+		// https://github.com/longhorn/longhorn/issues/7012
+		hasCorrespondingActiveReplica := func(replica *longhorn.Replica) bool {
+			for _, r := range rs {
+				if r.Spec.Active &&
+					r.Name != replica.Name &&
+					r.Spec.NodeID == replica.Spec.NodeID &&
+					r.Spec.DiskID == replica.Spec.DiskID &&
+					r.Spec.DiskPath == replica.Spec.DiskPath &&
+					r.Spec.DataDirectoryName == replica.Spec.DataDirectoryName {
+					return true
+				}
+			}
+			return false
+		}
+
+		// sort the replica by name so that we don't select multiple replicas when running this logic repeatedly quickly
+		var sortedReplicaNames []string
+		for rName := range rs {
+			sortedReplicaNames = append(sortedReplicaNames, rName)
+		}
+		sort.Strings(sortedReplicaNames)
+
+		var inactiveReplicaToCleanup *longhorn.Replica
+		for _, rName := range sortedReplicaNames {
+			r := rs[rName]
+			if !r.Spec.Active && !hasCorrespondingActiveReplica(r) {
+				inactiveReplicaToCleanup = r
+				break
+			}
+		}
+		if inactiveReplicaToCleanup != nil {
+			log.Infof("cleaning up inactive replica %v which doesn't have corresponding active replica", inactiveReplicaToCleanup.Name)
```

Comment:
NIT: I think our convention is to start log messages capitalized: https://github.com/longhorn/longhorn/wiki/coding-convention#wording.

Reply:
Thank you. Fixed